uamqp/aio/_client_async.py

```diff
@@ -407,3 +436,111 @@ async def receive_message_batch_async(self, max_batch_size=None, on_message_rece
+         goes into an error state.
+        """
+        # pylint: disable=protected-access
+        if not self._link:
+            self._link = self._session.create_sender_link(
+                target_address=self.target,
+                link_credit=self._link_credit,
+                send_settle_mode=self._send_settle_mode,
+                rcv_settle_mode=self._receive_settle_mode,
+                max_message_size=self._max_message_size,
+                properties=self._link_properties)
+            await self._link.attach()
+            return False
+        if self._link.state.value != 3:  # ATTACHED
+            return False
+        return True
+
+    async def _client_run_async(self, **kwargs):
+        """MessageSender Link is now open - perform message send
+        on all pending messages.
+        Will return True if operation successful and client can remain open for
+        further work.
+
+        :rtype: bool
+        """
+        try:
+            await self._connection.listen(**kwargs)
+        except ValueError:
+            _logger.info("Timeout reached, closing sender.")
+            self._shutdown = True
+            return False
+        return True
+
+    async def _transfer_message_async(self, message_delivery, timeout=0):
+        message_delivery.state = MessageDeliveryState.WaitingForSendAck
+        on_send_complete = partial(self._on_send_complete_async, message_delivery)
+        delivery = await self._link.send_transfer(
+            message_delivery.message,
+            on_send_complete=on_send_complete,
+            timeout=timeout
+        )
+        if not delivery.sent:
+            raise RuntimeError("Message is not sent.")
+
+    async def _on_send_complete_async(self, message_delivery, message, reason, state):
+        # TODO: check whether the callback would be called in case of message expiry or link going down
+        #  and if so handle the state in the callback
+        if state and SEND_DISPOSITION_ACCEPT in state:
+            message_delivery.state = MessageDeliveryState.Ok
+        else:
+            # TODO:
+            #  - sending disposition state could only be rejected/accepted?
+            #  - error action
+            #  - whether the state should be None in the case of sending failure/we could give a better default value
+            #   (message is not delivered)
+            if not state and reason == LinkDeliverySettleReason.NotDelivered:
+                message_delivery.state = MessageDeliveryState.Error
+                message_delivery.reason = reason
+                return
+
+            error_response = ErrorResponse(error_info=state[SEND_DISPOSITION_REJECT])
+            # TODO: check error_info structure
+            if error_response.condition == b'com.microsoft:server-busy':
+                # TODO: customized/configurable error handling logic
+                time.sleep(4)  # 4 is what we're doing nowadays in EH/SB, service tells client to backoff for 4 seconds
```

Comment:
I would rather not have us check for and act on 'microsoft' specific messages....

Reply:
this part of code has been updated to duplicate the sync implementation